client/src/components/hooks/useRequest.js

```diff
@@ -0,0 +1,52 @@
+import { useState, useEffect, useReducer } from 'react';
+
+const reducer = (state, action) => {
+  switch (action.type) {
+    case 'init':
+      return {
+        ...state,
+        isLoading: true,
+        isError: false,
+      };
+    case 'success':
+      return {
+        ...state,
+        isLoading: false,
+        isError: false,
+        data: action.data,
+      };
+    case 'fail':
+      return {
+        ...state,
+        isLoading: false,
+        isError: true,
+      };
+    default:
+      throw new Error();
+  }
+};
+
+const useRequest = (initialData, promise) => {
+  const [state, dispatch] = useReducer(reducer, {
+    data: initialData,
+    isLoading: true,
+    isError: null,
+  });
+
+  const [request, setRequest] = useState(() => promise);
+
+  useEffect(() => {
+    let cancelled = false;
+    dispatch({ type: 'init' });
+
+    request()
+      .then((data) => !cancelled && dispatch({ type: 'success', data }))
```

Comment:
usually types of actions are constants, so it should be 'SUCCESS'

Reply:
resolved